torchtext/csrc/core/dictionary.cpp

```diff
@@ -0,0 +1,45 @@
+#include "dictionary.h"
+#include <iostream>
+
+namespace torch {
+namespace text {
+namespace core {
+size_t Dictionary::word_hash(const std::string& word) {
+  size_t hash = 2166136261;
+  for (auto C : word) {
+    hash = hash ^ uint32_t(uint8_t(C));
+    hash = hash * 16777619;
+  }
+
+  return hash;
+}
+
+size_t Dictionary::find(const std::string& word) {
+  auto words_size = words.size();
+  auto index = word_hash(word) % words_size;
+
+  while (words[index].index != negative && words[index].word != word)
+    index = (index + 1) & words_size;
+
+  return index;
+}
+
+size_t Dictionary::add_word(const std::string& word) {
```

Comment:
After adding words, at some point, the user will ideally want to sort the dataset based on the frequency. The most frequent word should be the index 0, the second most frequent index 1, etc. - This is a memory optimization. Words follow [Zipf's law](https://en.wikipedia.org/wiki/Zipf%27s_law). When you read in a cacheline of data from this lookup you're most likely going to read index 0. Since you always read 64 bytes, that'll result in certain number of the first few words to always be read. The probability of hitting an index close to 0 is very high due to Zipf's law, so on the second read you'll likely be able to use the data you just read into cache. If the indicies on the other hand are random you'll constantly trash cache for a single word and read much more than you need to.

On that note, using std::string is probably not a good idea, because you need tight control over the structure that's used to represent the word. We can however add these optimizations later on once you have a benchmark (please add that as well) and can look at the memory pressure of a typical read. A good benchmark is tokenizing and creating a dictionary based on Wikipedia. At some point we might also want to align an entry struct or use SmallVector to store the first few characters on the stack.